packages/loader/container-utils/src/liveSummarySanitizer.ts

```diff
@@ -0,0 +1,194 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    ICommittedProposal,
+    ISummaryTree,
+    SummaryType,
+    ISummaryBlob,
+} from "@fluidframework/protocol-definitions";
+
+import { assert, bufferToString } from "@fluidframework/common-utils";
+
+function readBlobContent(content: ISummaryBlob["content"]): any {
+    const json =
+        typeof content === "string" ? content : bufferToString(content, "utf8");
+    return JSON.parse(json);
+}
+
+function validateKeyPresence(
+    keys: string[],
+    srcTree: ISummaryTree,
+): boolean {
+    const filtered = Object.keys(srcTree.tree).filter((i) =>
+        keys.includes(i),
+    );
+    return filtered.length === keys.length;
+}
+
+function createBlobItem(content: any): ISummaryBlob {
+    return {
+        content: JSON.stringify(content),
+        type: SummaryType.Blob,
+    };
+}
+
+function validateTree(liveSummary: ISummaryTree): boolean {
+    for (const key of Object.keys(liveSummary.tree)) {
+        const o = liveSummary.tree[key];
+        if (o.type === SummaryType.Handle) {
+            return false;
+        } else if (o.type === SummaryType.Tree && !validateTree(o)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+class LiveV1SummarySanitizer {
+    private readonly quorumKeyIdx = 0;
+    private readonly quorumValIdx = 1;
+    private readonly emptySummaryBlob: ISummaryBlob = {
+        content: JSON.stringify([]),
+        type: SummaryType.Blob,
+    };
+
+    constructor(private readonly srcSummary: ISummaryTree) {}
+
+    public get sanitizedCopy(): ISummaryTree {
+        const appTree = this.sanitizeAppTree(this.srcSummary);
+        const protocolTree = this.sanitizeProtocolTree(this.srcSummary);
+        return {
+            type: SummaryType.Tree,
+            tree: {
+                ".app": appTree,
+                ".protocol": protocolTree,
+            },
+        };
+    }
+
+    /**
+     * Sanitize app tree
+     * @param srcSummary - summary to transform
+     */
+    private sanitizeAppTree(srcSummary: ISummaryTree): ISummaryTree {
+        const keys = [".channels", ".electedSummarizer", ".metadata"];
```

Comment:
All of these terms (like ".channels") are about specific implementation (ContainerRuntime class). Any other implementation of container / IRuntime interface may have completely different content / representation.

These terms should not show up under packages/loader path. The only place they can be used is under packages/runtime/container-runtime

Reply:
Also, what about the other blobs / trees added at container runtime like ".blobs", ".chunks", ".alias", etc.?

Reply:
I can move this to runtime-utils or similar place. 

@agarwal-navin The recovery here is around v1 format. If by "keeping in sync" means making sure that changes in runtime schema should be reflected here, we should solve this through versioning. This logic should know how to sanitize v1 schema, nothing else.

Reply:
I personally do not like notion of V1 in this context. This is loader layer, and this layer has no relationship to container data schema versioning. All it knows is protocol tree structure, that's it. It does not care what container is opened and its schema.
Moreover, versioning of this layer is orthogonal to versioning of runtime layer, I.e., today's version of loader needs to continue to work seamlessly with a version of runtime that ships in a year.
There is nothing to "keep in sync", loader by design is decoupled from runtime.

I'd add that even if this code is in runtime, it would still be very hard (and likely wrong) to have that kind of knowledge / coupling with various features in runtime. For example, some things have feature gates that control file format (.channels), there is active work happening (new features like aliases, blobs) that may change file formats.

Reply:
Maybe there are two topics here:

1. This should not live at loader level. Maybe it was not clear from my comments but - agree. The location of this utils file should be on the runtime side, as it deals with runtime concerns.
2. We could reduce coupling by consuming some of the definitions from `summaryFormats` (runtime), so whether it's runtime or sanitizer, we are all landing on same definitions and/or setup functions. Or third option, expose new `sanitizeSummary` API on runtime itself and fully let the runtime deal with transformation, feature toggle handling and back-compat (against older versions). The last two items, runtime is doing already.

Reply:
Now if this logic ends up residing on the runtime side, it probably only makes sense to consume it through container runtime itself.

Reply:
`getSanitizedCopy` was removed. We are moving to option (3) as discussed here: https://github.com/microsoft/FluidFramework/issues/9442.